File: common/redis-cache/src/main/java/com/halcyon/tinder/rediscache/CacheManager.java
package com.halcyon.tinder.rediscache;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.time.Duration;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class CacheManager {

    private final RedisTemplate<String, Object> redisTemplate;

    public void save(String key, Object value, Duration timeToLive) {
        redisTemplate.opsForValue().set(key, value, timeToLive);
        log.info("Cached value with key '{}' for {} seconds", key, timeToLive.toSeconds());
    }

    public void save(String key, Duration timeToLive) {
        redisTemplate.opsForValue().set(key, "", timeToLive);
        log.info("Cached non with key '{}' for {} seconds", key, timeToLive.toSeconds());
    }

    public boolean isPresent(String key) {
        Object fetchedValue = redisTemplate.opsForValue().get(key);
        return Optional.ofNullable(fetchedValue).isPresent();
    }

    public <T> Optional<T> fetch(String key, Class<T> targetClass) {
        Optional<Object> value = Optional.ofNullable(redisTemplate.opsForValue().get(key));

        if (value.isEmpty()) {
            log.info("No cached value found for key '{}'", key);
            return Optional.empty();
        }

        var objectMapper = new ObjectMapper();
        objectMapper.registerModule(new JavaTimeModule());

        T result = objectMapper.convertValue(value.get(), targetClass);
        log.info("Fetched cached value with key {}", key);

        return Optional.of(result);
    }
}

File: common/redis-cache/src/main/kotlin/com/halcyon/tinder/rediscache/CacheManager.kt
package com.halcyon.tinder.rediscache

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Component
import java.time.Duration

@Component
class CacheManager(
    private val redisTemplate: RedisTemplate<String, Any>,
    private val objectMapper: ObjectMapper
) {

    private val log = LoggerFactory.getLogger(CacheManager::class.java)

    fun save(key: String, value: Any, timeToLive: Duration) {
        redisTemplate.opsForValue().set(key, value, timeToLive)
        log.info("Cached value with key '{}' for '{}'", key, timeToLive.seconds)
    }

    fun save(key: String, timeToLive: Duration) {
        redisTemplate.opsForValue().set(key, "", timeToLive)
        log.info("Cached non with key '{}' for '{}'", key, timeToLive)
    }

    fun isPresent(key: String): Boolean {
        return redisTemplate.opsForValue().get(key) != null
    }

    fun <T> fetch(key: String, targetClass: Class<T>): T? {
        val value = redisTemplate.opsForValue().get(key) ?: return null
        return try {
            objectMapper.convertValue(value, targetClass)
        } catch (e: Exception) {
            log.error("Failed to convert cached value for key '{}'", key, e)
            null
        }
    }

    fun delete(key: String) {
        redisTemplate.delete(key)
    }
}